lib/IRGen/GenDecl.cpp

```diff
@@ -802,8 +802,19 @@ void IRGenerator::emitGlobalTopLevel() {
   bool isWholeModule = PrimaryIGM->getSILModule().isWholeModule();
   for (SILFunction &f : PrimaryIGM->getSILModule()) {
     // Only eagerly emit functions that are externally visible.
-    if (!isPossiblyUsedExternally(f.getLinkage(), isWholeModule))
+    if (!isPossiblyUsedExternally(f.getLinkage(), isWholeModule)) {
+      // This function is not externally visible.
+      // It could be still useful to emit it for debug purposes if it is
+      // a debug build with disabled optimizations.
+      // Do not emit this function if it is coming from a binary SIL file
+      // or if debug info is not supposed to be emitted or if it is a JIT mode.
+      if (!PrimaryIGM->IRGen.Opts.Optimize && !PrimaryIGM->IRGen.Opts.UseJIT &&
+          f.hasLocation() && !f.getLocation().isSILFile() &&
+          PrimaryIGM->DebugInfo) {
```

Comment:
In all of LLVM it is considered a bug if the debug options affect codegen. IMHO, we should also try to live by this rule in Swift. Practically, there are few reasons to build at -Onone without debug information anyway, so I think we don't loose much by just deciding based on IRGen.Opts.Optimize.